### PR TITLE
Fix/tokenization with preprocessing

### PR DESCRIPTION
**Pull request by kuppulur**
- Our current model pipeline doesn't handle newline characters like ("\n", "\r") and multiple space characters.
- In this PR, I added a preprocessing routine to remove any such characters in the problem descriptions and preprocess the problem text before extracting entities.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,6 @@
 # Changelog
 
-| Date        | Notes                               |
-|-------------|-------------------------------------|
-| May 3, 2023 | Initial release of Ner4Opt library. |
+| Date         | Notes                               |
+|------------- |-------------------------------------|
+| Mar 11, 2024 | Add a new preprocessing routine before extracting entites. |
```
Comment (skadio) [inline CHANGELOG.md:5]: you can add Issue link here to make it explict and reveal the context

File: ner4opt/ner4opt.py [hunk 2]
```diff
@@ -146,6 +148,9 @@ def get_entities(self, text: str) -> List[dict]:
         if not text:
             return predicted_entities
 
+        # preprocess
+        text = preprocess(text)
```
Comment (skadio) [inline ner4opt/ner4opt.py:152]: learning q -- how does this work with or relate to our L2 augmentation? That happens later on? And what happens to L2 when it cannot find an objective min/max?  OR.. it does not care about that, simply grabs last 2 regardless?

**Review by skadio — APPROVED** [review]
LGTM

File: ner4opt/utils.py [hunk 3]
```diff
@@ -19,6 +19,13 @@
 pluralizer = inflect.engine()
 
 
+def preprocess(text: str) -> str:
```
Comment (skadio) [inline ner4opt/utils.py:22]: so this does not break anything from the past? 

Also, please add a few unit tests. Add a unit test for the actual issue, drop a comment on issue number and say that this unit test comes from that issue. Add other unit test to show that this new preprocess() works, and a few tests for edge cases, empty input etc, if not tested already

**Review by skadio — APPROVED** [review]

File: ner4opt/utils.py (hunk 3 above)
Comment (kuppulur) [reply ner4opt/utils.py:22]: Yes, I checked if all our previous tests are passing, they are. 

Good idea, I will add a test for this too.

File: ner4opt/ner4opt.py (hunk 2 above)
Comment (kuppulur) [reply ner4opt/ner4opt.py:152]: Yes, l2_augment is not affected by this. We augment regardless of whether we find an objective name or not. We assume that it is a fundamental property of pretty much all our problems, at least for our case (training data and dev data)

File: CHANGELOG.md (hunk 1 above)
Comment (kuppulur) [reply CHANGELOG.md:5]: Sounds good, will do that.

Comment (skadio) [reply CHANGELOG.md]: plz also thank the issue submitter using @ sign

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,5 +1,6 @@
 # Changelog
 
-| Date        | Notes                               |
-|-------------|-------------------------------------|
-| May 3, 2023 | Initial release of Ner4Opt library. |
+| Date         | Notes                               |
+|------------- |-------------------------------------|
+| Mar 11, 2024 | Add a new preprocessing routine before extracting entites. https://github.com/skadio/ner4opt/issues/4|
```
Comment (skadio) [inline CHANGELOG.md:5]: add a quick thank you to the submitter

File: tests/test_utils.py [hunk 5]
```diff
@@ -18,6 +18,11 @@ def setUp(self):
         self._example_for_spacy_tokenization = "Company A has a profit margin of 10% and B has 20%."
         self._words_and_phrases = ["batch", "coconut", "actor", "wheel", "blood pressure medicine"]
         self._spacy_sentence = nlp('Let us learn about " puff of air test " and a sewing - machine')
+        self._unnormalized_text = "  This\nis\r\na\nstring\n\rwith\n\rmultiple\n\r\nnewlines\n\n\rand   spaces.  "
```
Comment (skadio) [inline tests/test_utils.py:21]: do you mean raw_text?

Comment (skadio) [reply tests/test_utils.py:21]: what happens for empty "" text? Do we have a unit test for that?

**Review by skadio — APPROVED** [review]

**Review by skadio — APPROVED** [review]

File: tests/test_utils.py (hunk 5 above)
Comment (kuppulur) [reply tests/test_utils.py]: We have an empty string test here https://github.com/skadio/ner4opt/blob/66820b051bf89087f82602c2d2b004b944ddd5eb/tests/test_ner4opt.py#L78
